### PR TITLE
RPD unwrench upgrade

### PR DESCRIPTION
**Pull request by Dennok**
## About The Pull Request

Adds reverse wrench mode upgrade to the RPD.
Attention, due to budget cuts, the mode is hard linked to the destroy mode control button.
finished #53607

Now `/obj/machinery/atmospherics/deconstruct(disassembled = TRUE)` return created `obj/item/pipe`

## Why It's Good For The Game

OP atmostech is good

## Changelog
:cl:
add: Reverse wrench mode upgrade to the RPD. Attention, due to budget cuts, the mode is hard linked to the destroy mode control button.
/:cl:

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Would this be something you'd actually want? If I hit a pipe with an rcd I'd want to place a new pipe, not unwrench one right?

**Comment by Dennok** [comment]
> Would this be something you'd actually want? If I hit a pipe with an rcd I'd want to place a new pipe, not unwrench one right?

 If you hit a pipe with an rcd you place unwrenched fitting over it, 
or paint it, if paint mode enabled

File: code/game/objects/items/RPD.dm [hunk 1]
```diff
@@ -212,6 +212,8 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	var/static/datum/pipe_info/first_disposal
 	var/static/datum/pipe_info/first_transit
 	var/mode = BUILD_MODE | DESTROY_MODE | WRENCH_MODE
+	///Hold upgrade flags
+	var/upgrade
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/RPD.dm:216]: please denote it uses flags in the varname `upgrade_flags` for example

File: code/game/objects/items/RPD.dm [hunk 2]
```diff
@@ -235,6 +237,16 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
 
+/obj/item/pipe_dispenser/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/rpd_upgrade))
+		var/obj/item/rpd_upgrade/rpd_up = W
+		if(!(upgrade & rpd_up.upgrade))
+			upgrade |= rpd_up.upgrade
+			playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+			qdel(W)
+	else
+		return ..()
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/RPD.dm:248]:
```DM
if(!thing)
    return ..()
code

File: code/game/objects/items/RPD.dm [hunk 3]
```diff
@@ -338,6 +350,11 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	var/queued_p_dir = p_dir
 	var/queued_p_flipped = p_flipped
 
+	//Unwrench pipe before we build one over/paint it.
+	if((mode & DESTROY_MODE) && (upgrade & RPD_UPGRADE_UNWRENCH))
+		if(istype(A, /obj/machinery/atmospherics))
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/RPD.dm:355]:
```DM
	if((mode & DESTROY_MODE) && (upgrade & RPD_UPGRADE_UNWRENCH) && istype(A, /obj/machinery/atmospherics))
```

File: code/modules/research/techweb/all_nodes.dm [hunk 4]
```diff
@@ -528,9 +528,9 @@
 
 /datum/techweb_node/rcd_upgrade
 	id = "rcd_upgrade"
-	display_name = "RCD designs upgrade"
-	description = "Unlocks new RCD designs."
-	design_ids = list("rcd_upgrade_frames", "rcd_upgrade_simple_circuits", "rcd_upgrade_furnishing")
+	display_name = "RCD/RPD designs upgrade"
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/research/techweb/all_nodes.dm:531]:
```suggestion
	display_name = "rapid device upgrade designs"
```

File: code/modules/research/techweb/all_nodes.dm [hunk 5]
```diff
@@ -528,9 +528,9 @@
...
+	description = "Unlocks new RCD/RPD designs."
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/research/techweb/all_nodes.dm:532]:
```suggestion
	description = "Unlocks new designs that improve rapid devices."
```

**Comment by LemonInTheDark** [comment]
Tested, clean as fuck.

File: code/game/objects/items/RPD.dm [hunk 6]
```diff
@@ -338,6 +350,11 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+			A = A.wrench_act(user, src)
```
Comment (Dennok) [inline code/game/objects/items/RPD.dm:356]:
```suggestion
	if((mode & DESTROY_MODE) && (upgrade & RPD_UPGRADE_UNWRENCH) && istype(A, /obj/machinery/atmospherics))
		A = A.wrench_act(user, src)	
```

File: code/game/objects/items/RPD.dm [hunk 7]
```diff
@@ -235,6 +237,16 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+/obj/item/pipe_dispenser/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/rpd_upgrade))
+		var/obj/item/rpd_upgrade/rpd_up = W
+		if(!(upgrade & rpd_up.upgrade))
+			upgrade |= rpd_up.upgrade
+			playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+			qdel(W)
+	else
+		return ..()
+
```
Comment (Dennok) [inline code/game/objects/items/RPD.dm:249]:
```suggestion
/obj/item/pipe_dispenser/attackby(obj/item/W, mob/user, params)
	if(!istype(W, /obj/item/rpd_upgrade))
		return ..()	
	var/obj/item/rpd_upgrade/rpd_up = W
	if(!(upgrade & rpd_up.upgrade))
		upgrade |= rpd_up.upgrade
		playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
		qdel(W)

```

File: code/game/objects/items/RPD.dm [hunk 8]
```diff
@@ -212,6 +212,8 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	var/static/datum/pipe_info/first_disposal
 	var/static/datum/pipe_info/first_transit
 	var/mode = BUILD_MODE | DESTROY_MODE | WRENCH_MODE
+	///Hold upgrade flags
+	var/upgrade_flags
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/RPD.dm:216]: doc

File: code/game/objects/items/RPD.dm [hunk 9]
```diff
@@ -482,3 +497,14 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 #undef DESTROY_MODE
 #undef PAINT_MODE
 #undef WRENCH_MODE
+
+/obj/item/rpd_upgrade
+	name = "RPD advanced design disk"
+	desc = "It seems to be empty."
+	icon = 'icons/obj/module.dmi'
+	icon_state = "datadisk3"
+	var/upgrade_flags
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/RPD.dm:506]: doc

**Review by ExcessiveUseOfCobblestone — COMMENTED** [review]
You have generated new variables/procs/defines without proper documentation for the autodoc to work.

Please click [here ](https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md)for a guide on how to properly document these items.

**Comment by ExcessiveUseOfCobblestone** [comment]
ping me on discord or something when this is done so i can merge

File: code/game/objects/items/RPD.dm (hunk 9 above)
Comment (Dennok) [reply code/game/objects/items/RPD.dm:506]:
```suggestion
	/// Bitflags for upgrades
	var/upgrade_flags
```

File: code/game/objects/items/RPD.dm [hunk 10]
```diff
@@ -235,6 +237,21 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+		install_upgrade(W, user)
+		return TRUE
+	return ..()
+
+/// Installs an upgrade into the RPD checking if it is already installed
+/obj/item/pipe_dispenser/proc/install_upgrade(obj/item/rpd_upgrade/rpd_up, mob/user)
+	if(rpd_up.upgrade & upgrade_flags)
+		to_chat(user, "<span class='warning'>[src] has already installed this upgrade!</span>")
+		return
+	upgrade_flags |= rpd_up.upgrade
```
Comment (Dennok) [inline code/game/objects/items/RPD.dm:251]:
```suggestion
	if(rpd_up.upgrade_flags& upgrade_flags)
		to_chat(user, "<span class='warning'>[src] has already installed this upgrade!</span>")
		return
	upgrade_flags |= rpd_up.upgrade_flags
```

File: code/game/objects/items/RPD.dm [hunk 11]
```diff
@@ -235,6 +237,21 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 /obj/item/pipe_dispenser/attack_self(mob/user)
 	ui_interact(user)
 
+/obj/item/pipe_dispenser/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/rpd_upgrade))
+		install_upgrade(W, user)
+		return TRUE
+	return ..()
+
+/// Installs an upgrade into the RPD checking if it is already installed
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/RPD.dm:246]: https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md#documenting-a-proc

Comment (Dennok) [reply code/game/objects/items/RPD.dm:246]:
```suggestion
/**
  * Installs an upgrade into the RPD
  *
  *  Installs an upgrade into the RPD checking if it is already installed
  * Arguments:
  * * rpd_up - RPD upgrade
  * * user - mob that use upgrade on RPD
  */
```

File: code/game/objects/items/RPD.dm [hunk 12]
```diff
@@ -235,6 +237,28 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+/obj/item/pipe_dispenser/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/rpd_upgrade))
+		install_upgrade(W, user)
+		return TRUE
+	return ..()
+
+/**
+  * Installs an upgrade into the RPD
+  *
+  *  Installs an upgrade into the RPD checking if it is already installed
```
Comment (Dennok) [inline code/game/objects/items/RPD.dm:249]:
```suggestion
  * Installs an upgrade into the RPD checking if it is already installed
```

File: code/game/objects/items/RPD.dm (hunk 11 above)
Comment (Dennok) [reply code/game/objects/items/RPD.dm]: @ExcessiveUseOfCobblestone Am I do it wrong or do you need something else?

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 13]
```diff
@@ -247,6 +247,7 @@
 			if(!disassembled)
 				stored.obj_integrity = stored.max_integrity * 0.5
 			transfer_fingerprints_to(stored)
+			. = stored
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/machinery/atmosmachinery.dm:250]: Could you add some documentation that mentions wrench_act/deconstruct returning the pipe they touched? Aside from that this looks great

Comment (Dennok) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:251]: @LemonInTheDark I added some doc.

**Pull request merged**